src/components/ListComposable/types.ts

```diff
@@ -0,0 +1,148 @@
+import type {QAProps} from '../types';
+
+import type {modToHeight} from './constants';
+
+export type ListItemId = string;
+/**
+ * Default fallback id type.
+ * The core idea is to store original index and and nested children index
+ */
+export type GroupedId = string;
+
+export interface ListItemType<T = unknown> {
+    /**
+     * Required if you want to control list state from internal place.
+     * For example to control expanded groups state
+     */
+    id?: string;
+    data: T;
+    children?: ListItemType<T>[];
+}
+export type GroupParsedState = {
+    childrenCount: number;
+};
+
+export type ItemParsedState = {
+    indentation: number;
+};
+export type ParsedState<T = unknown> = {
+    /**
+     * stored internal meta info about item
+     * Note: Groups are also items
+     */
+    itemsState: Record<ListItemId, ItemParsedState>;
+    /**
+     * Normalized original data
+     */
+    byId: Record<ListItemId, T>;
+    /**
+     * Re
+     */
+    groupsState: Record<ListItemId, GroupParsedState>;
+};
+
+export type ListSizeTypes = keyof typeof modToHeight;
+
+export interface ListItemBaseData {
+    title: string;
+    subtitle?: string;
+    leftSlot?: React.ReactNode;
+    rightSlot?: React.ReactNode;
+}
+
+export interface ListItemRendererProps<T> {
+    id: number;
+    item: T;
+}
+
+export interface ListContainerRenderProps<T> {
+    items: T[];
+    getItemSize(index: number): number;
+    className?: string;
+    children(props: ListItemRendererProps<T>): React.ReactNode;
+}
+
+export type OnListItemClick<T> = (
+    item: ListItemType<T>['data'],
+    itemId: ListItemId,
+    fromKeyboard?: boolean,
+) => void;
+
+export type SetActiveItem = (id: ListItemId | null) => void;
+
+export type ExpandedState = Record<ListItemId, boolean>;
+
+export type SetGroupState = (value: (currentState: ExpandedState) => ExpandedState) => void;
+
+export interface RenderListItemViewProps extends QAProps {
+    /**
+     * @default `m`
+     */
+    size: ListSizeTypes;
+    isGroup: boolean;
+    selected: boolean;
+    active: boolean;
+    disabled: boolean;
+    /**
+     * By default hovered elements has active styles. You can switch off this behavior
+     */
+    activeOnHover: boolean;
+    indentation: number;
+    onClick?(): void;
+    title: string | React.ReactNode;
+    subtitle?: string;
+    leftSlot: React.ReactNode;
+    rightSlot: React.ReactNode;
+}
+
+export type ListContextType<T> = ParsedState<T> & {
+    activeItem: ListItemId | null;
+    size: ListSizeTypes;
+    order: ListItemId[];
+    handleKeyDown(e: React.KeyboardEvent<HTMLUListElement | HTMLDivElement>): void;
```

Comment:
`onKeyDown`?

Reply:
![image](https://github.com/gravity-ui/uikit/assets/27916444/3f085f77-d9c5-492a-8fd3-454e35d7e6a3)
Old list has this commonet

Reply:
![image](https://github.com/gravity-ui/uikit/assets/27916444/3f085f77-d9c5-492a-8fd3-454e35d7e6a3)
Old list has this comoment